dlt/pipeline/helpers.py

```diff
@@ -32,4 +36,100 @@ def _retry_load(ex: Exception) -> bool:
             return False
         return True
 
-    return _retry_load
\ No newline at end of file
+    return _retry_load
+
+
+class _DropCommand:
+    def __init__(
+        self,
+        pipeline: Pipeline,
+        resources: Optional[Union[Iterable[str], str]] = None,
+        schema_name: str = None,
+        state_paths: Optional[Union[Iterable[str], str]] = None
+    ) -> None:
+        self.pipeline = pipeline
+        # self.tables = set(tables or [])
+        if isinstance(resources, str):
+            resources = [resources]
+        if isinstance(state_paths, str):
+            state_paths = [state_paths]
+        resources = set(resources or [])
+        self.schema = pipeline.schemas[schema_name or pipeline.default_schema_name].clone()
+        self.schema_tables = self.schema.tables
+
+        drop_tables = []
+        for tbl in self.schema_tables.values():
+            if tbl.get('resource') in resources:
+                drop_tables += get_child_tables(self.schema_tables, tbl['name'])
+        self.drop_tables = list(reversed(drop_tables))
+
+        # # List resource state keys to drop
+        # drop_resources = set(r for r in (t.get('resource') for t in self.drop_tables) if r)
+        self.drop_state_paths: List[JSONPath] = [jsonpath_parse(p) for p in state_paths or []]
+        self.drop_resource_state_keys = resources
+
+    def drop_destination_tables(self) -> None:
+        with self.pipeline._get_destination_client(self.schema) as client:
+            client.drop_tables(*[tbl['name'] for tbl in self.drop_tables])
+
+    def delete_pipeline_tables(self) -> None:
+        for tbl in self.drop_tables:
+            del self.schema_tables[tbl['name']]
+        self.schema.bump_version()
+
+    def _list_state_paths(self, source_state: Dict[str, Any]) -> Iterator[str]:
+        for path in self.drop_state_paths:
+            matches = path.find(source_state)
+            for match in matches:
+                yield str(match.full_path)
+
+    def _drop_state_paths(self, source_state: Dict[str, Any]) -> None:
+        for path in self.drop_state_paths:
+            path.filter(lambda _: True, source_state)
+
+    def drop_state_keys(self) -> None:
+        state: TSourceState
+        with self.pipeline.managed_state(extract_state=True, extract_unchanged=True) as state:  # type: ignore[assignment]
+            source_states = state.get("sources")
+            if not source_states:
+                return
+            # TODO: Should only drop from source state matching the schema. Not possible atm
+            for source_state in source_states.values():
+                for p in self.drop_state_paths:  # Remove keys by JSON path
+                    p.filter(lambda _: True, source_state)
+                resources = source_state.get('resources')
```

Comment:
this should be a part of `common/pipeline.py` there's a function that resets the state there. maybe you can use it. if not please keep all operations that rely on our state layout in there. I refer to you getting resources here and assuming it has a list of sources